### ISSUE TITLE
panic: interface conversion: interface {} is nil, not string

### ISSUE THREAD

**Issue by nishkrishnan** [hashicorp/terraform-provider-aws#22959]
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Terraform CLI and Terraform AWS Provider Version

Terraform version 1.0.1
Terraform AWS Provider version: 3.74.0

### Affected Resource(s)

* aws_iam_policy_document

### Terraform Configuration Files

```hcl

variable "additional_policy_statements" {
  type        = string
  description = "additional policy statements that can be added to the role's policy document"
  default     = ""
}

data "aws_iam_policy_document" "assume_role_policy" {
  source_policy_documents = [
    data.aws_iam_policy_document.partial.json,
    var.additional_policy_statements
  ]
}

data "aws_iam_policy_document" "partial" {
  statement {
    actions = ["sts:AssumeRole"]

    principals {
      type        = "Service"
      identifiers = ["ec2.amazonaws.com"]
    }
  }
}

resource "aws_iam_role" "role" {
  name               = "foo"
  assume_role_policy = data.aws_iam_policy_document.assume_role_policy.json
}

```

### Panic Output

```
Error: Plugin did not respond
│ 
│   with module.service_iam_role.data.aws_iam_policy_document.assume_role_policy,
│   on ../../../iam_role/assumable_roles.tf line 41, in data "aws_iam_policy_document" "assume_role_policy":
│   41: data "aws_iam_policy_document" "assume_role_policy" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ReadDataSource call. The plugin logs may contain
│ more details.
╵

Stack trace from the terraform-provider-aws_v3.74.0_x5 plugin:

panic: interface conversion: interface {} is nil, not string

goroutine 273 [running]:
github.com/hashicorp/terraform-provider-aws/internal/service/iam.dataSourcePolicyDocumentRead(0xc001c21880, 0x8268060, 0xc001170000, 0xf0f1910, 0xf0c30e0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/src/github.com/hashicorp/terraform-provider-aws/internal/service/iam/policy_document_data_source.go:138 +0x1a7f
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).read(0xc0007e9a40, 0xa507158, 0xc00203b3c0, 0xc001c21880, 0x8268060, 0xc001170000, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:346 +0x1ee
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*Resource).ReadDataApply(0xc0007e9a40, 0xa507158, 0xc00203b3c0, 0xc001c21780, 0x8268060, 0xc001170000, 0xc001170000, 0xc001c21780, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/resource.go:569 +0xfd
github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema.(*GRPCProviderServer).ReadDataSource(0xc000311038, 0xa507158, 0xc00203b3c0, 0xc001539360, 0x95a66e1, 0x12, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-sdk/v2@v2.10.1/helper/schema/grpc_provider.go:1133 +0x518
github.com/hashicorp/terraform-plugin-go/tfprotov5/tf5server.(*server).ReadDataSource(0xc0001a4a80, 0xa507200, 0xc00203b3c0, 0xc002268aa0, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/tf5server/server.go:478 +0x2f5
github.com/hashicorp/terraform-plugin-go/tfprotov5/internal/tfplugin5._Provider_ReadDataSource_Handler(0x930e3e0, 0xc0001a4a80, 0xa507200, 0xc00202ef90, 0xc0024d9da0, 0x0, 0xa507200, 0xc00202ef90, 0xc00005a900, 0x2c6)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/github.com/hashicorp/terraform-plugin-go@v0.5.0/tfprotov5/internal/tfplugin5/tfplugin5_grpc.pb.go:416 +0x214
google.golang.org/grpc.(*Server).processUnaryRPC(0xc00032dc00, 0xa533c18, 0xc001540600, 0xc0002a6500, 0xc0016e5320, 0xf0b1c50, 0x0, 0x0, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1194 +0x52b
google.golang.org/grpc.(*Server).handleStream(0xc00032dc00, 0xa533c18, 0xc001540600, 0xc0002a6500, 0x0)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:1517 +0xd0c
google.golang.org/grpc.(*Server).serveStreams.func1.2(0xc000552c90, 0xc00032dc00, 0xa533c18, 0xc001540600, 0xc0002a6500)
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:859 +0xab
created by google.golang.org/grpc.(*Server).serveStreams.func1
	/opt/teamcity-agent/work/5d79fe75d4460a2f/pkg/mod/google.golang.org/grpc@v1.32.0/server.go:857 +0x1fd

Error: The terraform-provider-aws_v3.74.0_x5 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.

```

### Expected Behavior

Empty string being passed in as a document should be gracefully handled.

### Actual Behavior

Panic.

### Steps to Reproduce

1. Use `aws_iam_policy_document` couple with `source_policy_documents`
2. Pass empty string as one of the source policy documents.
3. Run terraform init
4. Run terraform plan

**Comment by ewbankkit** [comment]
https://github.com/hashicorp/terraform-provider-aws/blob/e45aab12155799fd81bfb7ea66432dc0f5020e4a/internal/service/iam/policy_document_data_source.go#L126-L155

**Comment by jarrod-manwaring** [comment]
I ran into this today as well. The workaround I used was to wrap the list in a `compact()` function.
```hcl
data "aws_iam_policy_document" "assume_role_policy" {
  source_policy_documents = compact([
    data.aws_iam_policy_document.partial.json
    var.additional_policy_statements
  ])
}
```

**Comment by SteveRuble** [comment]
Possible simpler repro: just don't set the `policy` on a `aws_kms_key`. I consistently get this panic when I comment out the `policy` line.

**Comment by aviisingh** [comment]
Facing exact same issue. Terraform plan works fine but apply fails.
Error: Plugin did not respond
│ 
│   with module.default.module.cdn_endpoint_frontend_india.azurerm_cdn_endpoint.ce,
│   on main.tf line 8, in resource "azurerm_cdn_endpoint" "ce":
│    8: resource "azurerm_cdn_endpoint" "ce" {
│ 
│ The plugin encountered an error, and failed to respond to the
│ plugin.(*GRPCProvider).ApplyResourceChange call. The plugin logs may
│ contain more details.

Stack trace from the terraform-provider-azurerm_v2.99.0_x5 plugin:

panic: interface conversion: interface {} is nil, not string

**Comment by piotr-jagiello** [comment]
As an additional comment, I encountered it because source_json is now considered deprecated, and source_policy_documents is being presented as a replacement, but "source_json" happily accepts empty strings. (It doesn't say anywhere that source_policy_documents also does, but at the very least it shouldn't crash the plugin).

**Comment by carterjones** [comment]
For those with a single entry in `source_policy_documents`, something like this can help work around the bug for now by using conditional logic that only sets `source_policy_documents` if a variable is a non-empty string. Otherwise, it defaults to setting it to null, which does not make the bug manifest.

```terraform
data "aws_iam_policy_document" "assume_role_policy" {
  source_policy_documents = length(var.additional_policy_statements) > 0 ? [var.additional_policy_statements] : null
}
```

**Comment by YakDriver** [comment]
Looks like this was fixed with #27055. I added an additional test to ensure both `override_policy_documents` and `source_policy_documents` work with empty strings, ensuring there's no panic.
